### ISSUE TITLE
Examples needed

### ISSUE THREAD

**Issue by schw4rzlicht** [asyncapi/nodejs-ws-template#13]
Hey there!

I'm currently documenting a proprietary API using AsyncAPI and want to generate client code using the template.

I struggle with it because of the lack of websocket examples for AsyncAPI and how to use the generated code. What I learned so far is that there is only one "channel" when using web sockets with multiple messages to publish and subscribe to in most cases.

In my understanding, this template only generates one function to send messages for publish and subscribe sections. So building listeners for incoming messages and send functions for outgoing messages is still to be done manually, right? What is the whole point of generating node.js code then?

Is this template only meant as an example for how to build a template on your own which covers your needs?

Then again: I am not sure how to exactly describe an API using AsyncAPI in the first place. E.g. I do have a query parameter which is always present and never changes. But where would I define that? I tried bindings in the server object and bindings in the channel object but the only thing that seems to look right is putting it either into the server URL which then breaks the channel or using it as the channel name which breaks the generated code.

Could someone please shed some light into the dark? Thanks very much! :)

**Comment by fmvilas** [comment]
Hi @schw4rzlicht. I'm sorry to hear you're struggling with this. We're in the process of rebuilding our docs and adding more functionality to our code generators. This template was really an example and hasn't been taken care of much. That doesn't mean it's abandoned. We'll have to work on it soon and probably make it client only because we already have server-side generation for WS on github.com/asyncapi/nodejs-template.

> So building listeners for incoming messages and send functions for outgoing messages is still to be done manually, right? What is the whole point of generating node.js code then?

The purpose of this template is to have them. We just haven't worked on it yet. A PR would be appreciated 😊 

> Then again: I am not sure how to exactly describe an API using AsyncAPI in the first place. E.g. I do have a query parameter which is always present and never changes. But where would I define that? I tried bindings in the server object and bindings in the channel object but the only thing that seems to look right is putting it either into the server URL which then breaks the channel or using it as the channel name which breaks the generated code.

Channel bindings are the right place for this information: https://github.com/asyncapi/bindings/tree/master/websockets. Again, it just doesn't work because this template needs some love.

Here's an example using WebSockets: https://github.com/asyncapi/generator/blob/master/test/docs/ws.yml

I hope this helps. This is all we got so far in terms of documentation but stay tuned, we're working on great docs and guides as well as improving the templates.

**Comment by schw4rzlicht** [comment]
Thank you so much for your help!

Based on your information I started to work on the template with the goal to PR when I'm done.

It works out so far, the only challenge is to understand and use Nunjucks the right way. What's not clear to me is how resolving of messages and their names should work.

For channels I can easily do `{%- for channelName, channel in asyncapi.channels() %}` but not for messages since `asyncapi.channels()` returns an object while `channel.subscribe().messages()` returns an array, so I need to get the name (a.k.a. the key of the underlying YAML) of the message from every message object.

`message.name()` is empty; I could use `message._json.x-parser-message-name` for that but a) this looks bad and b) when printing it using `{{ }}` it only prints `NaN`. Is there a better way?

**Comment by aphilpott** [comment]
@schw4rzlicht I'm interested in using your fork, and potentially helping you get it to the point it could be merged into this repo (disclaimer: I am not a developer or maintainer of this repo).

Are you still working on it?

**Comment by schw4rzlicht** [comment]
@aphilpott Sorry for my late reply. Currently, I am not working on it any more due to the lack of feedback on the issue and changes in my field of work. Please feel free to use my fork and go from there, I might chime in at a later point again.

**Comment by derberg** [comment]
fyi @schw4rzlicht and @aphilpott, I will work on this topic in January (https://github.com/asyncapi/asyncapi/issues/253#issuecomment-726879745). Sorry we are so late here but the last 6months was a hell of a ride

**Comment by derberg** [comment]
still on my todo list, there was just a shift of priorities and because of AsyncAPI partnership with Postman we focused first on setting up open governance model and handing over the spec into the foundation. Sorry for the delay

**Comment by derberg** [comment]
@schw4rzlicht @aphilpott 
Hi folks. I recently fixed some bugs in the template and added the generation of some more stuff. The latest 0.9 release has those improvements. In the next 2 weeks, I'm gonna publish an article explaining what can be done with this template and the example that I used.

Bear in mind though that for now, this template doesn't support use case with single channels having `oneOf` multiple messages. It always reads first message from the list, so the use case is, multiple channels with single message each.
